### PR TITLE
feat: option to add private and public cert engines

### PR DESCRIPTION
**Pull request by jor2**
### Description

https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/issues/78

Expose ability to configure cert engine in sm instance

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

Expose ability to configure a public and private cert engine in a secrets manager instance.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:199]: no need to expose this - this can be determined based on the value used for the `allowed_network` variable.

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:184]: maybe use a better default value such as `default-cert-template` ?

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:183]: use proper english the the description. Don't use `private_cert`

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:177]: use proper english the the description. Don't use `private_cert`

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:162]: we probably need a better default value here, I'm just not sure what it would be. Other option is to default it to `null` but have logic to ensure its not null if `private_engine_enabled` is true. If we can find a good default value, I would prefer that, but it may be unique for every consumer?

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:129]: description is wrong - its talking about the IAM engine

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:113]: the DA deploys the SM instance, so you have this info already in the code

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:107]: the DA deploys the SM instance, so you have this info already in the code

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:101]: The DA creates the SM instance, so there won't be any secret ID available when deploying this DA. We could maybe take the ACME key value in as input and create the secret?

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:86]: this has no default value so its makes it a required input, however its would not be required if not deploying the cert engine right?

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:80]: descrption is wrong - its talking about the IAM credentials engine

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:73]: descrption is wrong - its talking about the IAM credentials engine

File: solutions/standard/variables.tf
Comment (ocofaigh) [inline solutions/standard/variables.tf:11]: don't expose this

File: solutions/standard/provider.tf
Comment (ocofaigh) [inline solutions/standard/provider.tf:12]: don't expose this

**Review by ocofaigh — CHANGES_REQUESTED** [review]
By adding variable validation you are duplicating what is already in the module itself, meaning we have to maintain in 2 places. If the validation is already done i the module, I don't suggest duplicating it in the DA variables.tf.

I'm also wondering if its now time to create a new repo for the DA since it deploys more than just an SM instance. Similar to what we did with [terraform-ibm-observability-da](https://github.com/terraform-ibm-modules/terraform-ibm-observability-da). It would make versioning and maintenance much easier. Lets discuss internally

File: solutions/standard/variables.tf
Comment (jor2) [reply solutions/standard/variables.tf:183]: just came from here https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager-private-cert/blob/060eef08db1a7ffb48b7969e36cd729702da08a1/examples/default/variables.tf#L25

**Comment by jor2** [comment]
exception:
````
 2024/04/18 10:18:58 Terraform apply | module.private_secret_engine[0].ibm_sm_private_certificate_configuration_template.certificate_template: Creation complete after 1s [id=us-south/ec9e2484-9656-46fb-9540-71a086f93981/default-cert-template]
 2024/04/18 10:18:58 Terraform apply | 
 2024/04/18 10:18:58 Terraform apply | Error: [ERROR] Error creating serviceID: You are not authorized to use this API. {
 2024/04/18 10:18:58 Terraform apply |     "StatusCode": 403,
 2024/04/18 10:18:58 Terraform apply |     "Headers": {
 2024/04/18 10:18:58 Terraform apply |         "Akamai-Grn": [
 2024/04/18 10:18:58 Terraform apply |             "0.d1764017.1713435533.233a0619"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Cache-Control": [
 2024/04/18 10:18:58 Terraform apply |             "no-cache, no-store, must-revalidate"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Content-Language": [
 2024/04/18 10:18:58 Terraform apply |             "en-US"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Content-Length": [
 2024/04/18 10:18:58 Terraform apply |             "452"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Content-Type": [
 2024/04/18 10:18:58 Terraform apply |             "application/json"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Date": [
 2024/04/18 10:18:58 Terraform apply |             "Thu, 18 Apr 2024 10:18:53 GMT"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Expires": [
 2024/04/18 10:18:58 Terraform apply |             "0"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Pragma": [
 2024/04/18 10:18:58 Terraform apply |             "no-cache"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Strict-Transport-Security": [
 2024/04/18 10:18:58 Terraform apply |             "max-age=31536000; includeSubDomains"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "Transaction-Id": [
 2024/04/18 10:18:58 Terraform apply |             "ZnA0ODk-56cf2b45d7d340e0bdf946dcfa5d89f3"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "X-Content-Type-Options": [
 2024/04/18 10:18:58 Terraform apply |             "nosniff"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "X-Correlation-Id": [
 2024/04/18 10:18:58 Terraform apply |             "ZnA0ODk-56cf2b45d7d340e0bdf946dcfa5d89f3"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "X-Proxy-Upstream-Service-Time": [
 2024/04/18 10:18:58 Terraform apply |             "107"
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "X-Request-Id": [
 2024/04/18 10:18:58 Terraform apply |             "05300314-79b0-4de3-8776-1d505d7a37c5"
 2024/04/18 10:18:58 Terraform apply |         ]
 2024/04/18 10:18:58 Terraform apply |     },
 2024/04/18 10:18:58 Terraform apply |     "Result": {
 2024/04/18 10:18:58 Terraform apply |         "errors": [
 2024/04/18 10:18:58 Terraform apply |             {
 2024/04/18 10:18:58 Terraform apply |                 "code": "forbidden",
 2024/04/18 10:18:58 Terraform apply |                 "details": "Reason: Access check for action 'iam-identity.serviceid.create' for subject 'IBMid-666000KAO3' for resource 'a/abac0df06b644a9cabc6e44f55b3880e' was denied due to lack of access policy. Outbound transaction id 'ZnA0ODk-56cf2b45d7d340e0bdf946dcfa5d89f3-000'.",
 2024/04/18 10:18:58 Terraform apply |                 "message": "You are not authorized to use this API.",
 2024/04/18 10:18:58 Terraform apply |                 "message_code": "BXNIM0513E"
 2024/04/18 10:18:58 Terraform apply |             }
 2024/04/18 10:18:58 Terraform apply |         ],
 2024/04/18 10:18:58 Terraform apply |         "status_code": 403,
 2024/04/18 10:18:58 Terraform apply |         "trace": "ZnA0ODk-56cf2b45d7d340e0bdf946dcfa5d89f3"
 2024/04/18 10:18:58 Terraform apply |     },
 2024/04/18 10:18:58 Terraform apply |     "RawResult": null
 2024/04/18 10:18:58 Terraform apply | }
 2024/04/18 10:18:58 Terraform apply | 
 2024/04/18 10:18:58 Terraform apply | 
 2024/04/18 10:18:58 Terraform apply |   with module.iam_secrets_engine[0].ibm_iam_service_id.iam_secret_generator,
 2024/04/18 10:18:58 Terraform apply |   on .terraform/modules/iam_secrets_engine/main.tf line 18, in resource "ibm_iam_service_id" "iam_secret_generator":
 2024/04/18 10:18:58 Terraform apply |   18: resource "ibm_iam_service_id" "iam_secret_generator" {
````

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@jor2 I left some comments. Please don't run the pipeline again until PR has been approved (we are hitting SM quota issues in our account)

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@jor2 A recent PR just added input variables into `ibm_catalog.json`. The reason we are doing this is so they show up in a specific order in catalog / projects.
So for all the new vars you added here, you will need to now add them to ibm_catalog.json. Add them in the order we want them to display (so pubic cert ones together and private cert ones together)

**Comment by jor2** [comment]
Unless [`acme_letsencrypt_private_key`](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/90/files#diff-009e252987a2a3c3ea49bb0f9b3dc403769755965b5656e8b0a210d3499c0d6cR116) is a valid secret it will fail. Is there a downside to generating a secret using the [acme tool](https://github.com/ibm-cloud-security/acme-account-creation-tool) and pasting it here?

File: solutions/standard/main.tf
Comment (ocofaigh) [reply solutions/standard/main.tf:106]: have reached out to codeowners of this module to confirm if this is expected

Comment (ocofaigh) [reply solutions/standard/main.tf:106]: confirmed, its expected. Create an issue in that repo to update the usage https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager-public-cert-engine/issues/75

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**
